net/gcoap: support DTLS

janosbrodbeck opened the pull request:
This PR is still WIP. But the core mechanics work. Feedback in every way is very, very welcome!
### Contribution description
This PR adds DTLS support in gcoap. It adds the following:

* example/gcoap (gcoap-cli) supports DTLS 
* DTLS session management to support more than one session, hold sessions as long as possible and to ensure that no session gets lost
* Unified `_sock_send` function to abstract udp/dtls away from gcoap
* So far no API changes to gcoap. Coap part itself is hardly touched.

But it's still not complete:
* Testing of observable resources
* Fulfill all CoAP DTLS RFC requirements (e.g. CoAP requests needs to be handled in the same epoch)
* Possible additions to the DTLS-API / add some utility functionality to handle a few things easier (e.g., `sock_udp_ep_t` <-> `sock_dtls_session_t`, function to get epoch from session)
* Adjust documentation
* maybe more

And known bug(s) remain:
* How is it possible to increase `DTLS_PEER_MAX` for tinyDTLS? I've seen no file where it is set except the Kconfig file. It appears only in the build-files. Whatever I set, it is always 1 for tinyDTLS.

At the end are some more in depth comments regarding session management and some unchanged parameters.

Thanks to @pokgak for his gcoap/dtls PR, that helped me.
### Testing procedure

It's currently only tested between native RIOT nodes and between RIOT and [coap-shell](https://github.com/tzolov/coap-shell).

For testing between two RIOT nodes: flash `example/gcoap` and send a CoAP get request from the client to the server:
`coap get <ip> 5684 /riot/board`

For testing between RIOT and coap-shell: flash `example/gcoap` on one device/native and launch the coap-shell. For native testing:
```
connect --uri coaps://[<ip>%<interface>] --identity Client_identity --secret secretPSK
get --path /riot/board
```
### Issues/PRs references
Issue #15517 needs to be resolved. For now I've included a workaround commit till it's resolved. Workaround is just for testing purposes.

Relates to #10897

### Comments
__Session management:__ Gcoap only stores memos for client requests on client-side and registered observers on server-side. This causes clients to block internal session slots in tinyDTLS if clients crash or sessions are simply not resolved by the clients - the server gets completely blocked if all session slots are occupied. And these sessions/udp endpoints are not stored in any way. Gcoap has to make sure to never lose sessions and to free up slots on its own if neccessary. This sadly costs memory - but I have not found a way around it. 

Gcoap now reuses a session if it's already established. If the last session slot is occupied a configurable timeout will start and closes a session after expiration. Then the oldest, in terms of last usage, session will be closed.

__Not changing sock_udp in parameters__: In the long term it would be nice to support DTLS and UDP connections simultaneously if desired. Therefore I did not want to change any socket parameters. For supporting multiple transport layer we should maybe think about a more unified socket/tl type in gcoap which supports more than DTLS/UDP (if needed) and is also supported by the gcoap API. But that is out of scope of this PR I guess.

miri64 commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -317,6 +338,9 @@ int gcoap_cli_cmd(int argc, char **argv)
         uint8_t open_reqs = gcoap_op_state();
 
         printf("CoAP server is listening on port %u\n", CONFIG_GCOAP_PORT);
+        #ifdef CONFIG_GCOAP_ENABLE_DTLS
+            printf("Connection secured with DTLS\n");
+        #endif
```
```suggestion
        if (IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)) {
            puts("Connection secured with DTLS\n");
        }
```

miri64 commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -23,13 +23,34 @@
 #include "debug.h"
 
+#ifdef CONFIG_GCOAP_ENABLE_DTLS
+    #include "tinydtls_keys.h"
+    #include "net/credman.h"
+
+    #define SOCK_DTLS_CLIENT_TAG (2)
+    static const uint8_t psk_id_0[] = PSK_DEFAULT_IDENTITY;
+    static const uint8_t psk_key_0[] = PSK_DEFAULT_KEY;
+
+    static const credman_credential_t credential = {
+        .type = CREDMAN_TYPE_PSK,
+        .tag = SOCK_DTLS_CLIENT_TAG,
+        .params = {
+            .psk = {
+                .key = { .s = psk_key_0, .len = sizeof(psk_key_0) - 1, },
+                .id = { .s = psk_id_0, .len = sizeof(psk_id_0) - 1, },
+            }
+        },
+    };
```
Please don't indent on pre-processor expressions. Not only is it not within our style (we don't indent guards either) AFAIK most compilers that don't like `#` prefixed with white-spaces.

janosbrodbeck replied:
There are parts in RIOT which are barely readable because of the missing indents IMHO.  But when even some compilers complain about that I'm instantly convinced. Not necessary to spot more of those. I'll fix it!

miri64 replied:
Maybe you don't even need those, when using `IS_ACTIVE()` ;-)

miri64 commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -110,6 +123,15 @@ static event_queue_t _queue;
 static uint8_t _listen_buf[CONFIG_GCOAP_PDU_BUF_SIZE];
 static sock_udp_t _sock_udp;
 
+#ifdef CONFIG_GCOAP_ENABLE_DTLS
+    #define SOCK_DTLS_CLIENT_TAG (2)
+    static sock_dtls_t _sock_dtls;
+    static kernel_pid_t _main_thread;
+
+    static event_timeout_t _dtls_session_free_up_tmout;
+    static event_callback_t _dtls_session_free_up_tmout_cb;
+#endif
```
Here too

miri64 replied:
I think you can even drop the preprocessor `#if` around here and use `if (IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS))` through-out the code. This way we get rid of a lot of pre-processor magic, which allows you to not hide your code from the compiler in any case.

miri64 commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -128,14 +150,111 @@ static void *_event_loop(void *arg)
+    }
+}
+
+#else
```
Can we get around such large `#if` blocks? Those can easily get out of hand.

janosbrodbeck replied:
I think with using `IS_ACTIVE()` I can really drop most (maybe even all) `#if' blocks.

miri64 reviewed (changes requested):
I know this is still WIP, but I think you can get rid of most `#ifdef` / `#if` foobar, when using `IS_ACTIVE()` and relying on the optimizer.

miri64 commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -317,6 +338,9 @@ int gcoap_cli_cmd(int argc, char **argv)
 
         printf("CoAP server is listening on port %u\n", CONFIG_GCOAP_PORT);
+        #ifdef CONFIG_GCOAP_ENABLE_DTLS
+            printf("Connection secured with DTLS\n");
```
Maybe also print a finger-print of the PSK here, so a user can compare?

miri64 commented on examples/gcoap/Makefile:
```diff
@@ -27,6 +27,22 @@ USEMODULE += shell
 USEMODULE += ps
 
+# DTLS configuration
+GCOAP_ENABLE_DTLS ?= 0
+ifeq (1,$(GCOAP_ENABLE_DTLS))
+	CFLAGS += -DCONFIG_GCOAP_ENABLE_DTLS=1
+	CFLAGS += -DCONFIG_GCOAP_DTLS_SESSIONS_MAX=1
+	# UDP Port to use (20220 is default for DTLS).
+	DTLS_PORT ?= 20220
+	CFLAGS += -DDTLS_DEFAULT_PORT=$(DTLS_PORT)
+
+	USEMODULE += gcoap_dtls
+	USEPKG += tinydtls
+	USEMODULE += tinydtls_sock_dtls
+	# tinydtls needs crypto secure PRNG
+	USEMODULE += prng_sha1prng
+endif
```
Some of the config stuff might need to be moved lower to work with Kconfig. @leandrolanzieri @akshaim can one of you have a look?

leandrolanzieri replied:
If those parameters are added also to Kconfig (which I would really recommend :) then yes, the configuration changes need to be moved bellow the inclusion of `Makefile.include`. That way one is able to check if Kconfig is being used.

miri64 commented on examples/gcoap/Makefile:
```diff
@@ -27,6 +27,22 @@ USEMODULE += shell
+ifeq (1,$(GCOAP_ENABLE_DTLS))
+	CFLAGS += -DCONFIG_GCOAP_ENABLE_DTLS=1
+	CFLAGS += -DCONFIG_GCOAP_DTLS_SESSIONS_MAX=1
+	# UDP Port to use (20220 is default for DTLS).
+	DTLS_PORT ?= 20220
+	CFLAGS += -DDTLS_DEFAULT_PORT=$(DTLS_PORT)
```
I think changing the `DTLS_DEFAULT_PORT` does not make sense here. One might also want to use other applications using DTLS. Rather, a `GCOAPS_DEFAULT_PORT` should be introduced.

akshaim commented on examples/gcoap/Makefile:
```diff
@@ -27,6 +27,22 @@ USEMODULE += shell
+# DTLS configuration
+GCOAP_ENABLE_DTLS ?= 0
+ifeq (1,$(GCOAP_ENABLE_DTLS))
+	CFLAGS += -DCONFIG_GCOAP_ENABLE_DTLS=1
```
Please add the same to [gcoap/Kconfig](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/application_layer/gcoap/Kconfig), as a `bool`. Macros in `CONFIG_` namespace are expected to be exposed via Kconfig.

janosbrodbeck replied:
Was not my intention to not support Kconfig, just did not address it at this point. Will address Kconfig with the next fixups :)

pokgak commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -150,6 +269,35 @@ static void _on_sock_evt(sock_udp_t *sock, sock_async_flags_t type, void *arg)
+    /* send application data */
+    res = sock_dtls_send(&_sock_dtls, &session, data, len,
+                            2*CONFIG_GCOAP_DTLS_HANDSHAKE_TIMEOUT_USEC);
+    if (!res) {
```
```suggestion
    if (res < 0) {
```

pokgak commented on sys/net/application_layer/gcoap/dtls_session_management.c:
```diff
@@ -0,0 +1,165 @@
+static int _find_session(const sock_udp_ep_t *ep, dsm_session_t **session);
+
+static mutex_t _lock;
+static dsm_session_t _sessions[CONFIG_GCOAP_DTLS_SESSIONS_MAX];
```
```suggestion
static dsm_session_t _sessions[DTLS_PEER_MAX];
```

janosbrodbeck replied:
The idea behind introducing `CONFIG_GCOAP_DTLS_SESSIONS_MAX` was that `DTLS_PEER_MAX` is tied to tinyDTLS. So in the future there could be other DTLS pkgs (e.g. wolfSSL) which possibly have a different variable for that. So supporting more DTLS pkgs would go quicker in the future. 

But `CONFIG_GCOAP_DTLS_SESSIONS_MAX` needs to be exact the same as the DTLS implementation uses internally (because as far as I understand, you cannot specify how many sessions are allowed per socket).

pokgak commented on sys/net/application_layer/gcoap/dtls_session_management.c:
```diff
@@ -0,0 +1,165 @@
+
+void dtls_session_management_init(void) {
+    mutex_init(&_lock);
+    _available_slots = CONFIG_GCOAP_DTLS_SESSIONS_MAX;
```
```suggestion
    _available_slots = DTLS_PEER_MAX;
```

pokgak commented on sys/net/application_layer/gcoap/dtls_session_management.c:
```diff
@@ -0,0 +1,165 @@
+    mutex_init(&_lock);
+    _available_slots = CONFIG_GCOAP_DTLS_SESSIONS_MAX;
+
+    for (uint8_t i=0; i < CONFIG_GCOAP_DTLS_SESSIONS_MAX; i++) {
```
```suggestion
    for (uint8_t i=0; i < DTLS_PEER_MAX; i++) {
```

pokgak reviewed (commented):
I don't see gcoap specific code in the `dtls_session_management` module other than reading the max peer config option. Maybe we could move it out of gcoap and make it an optional module for sock dtls instead?

pokgak commented:
> * How is it possible to increase `DTLS_PEER_MAX` for tinyDTLS? I've seen no file where it is set except the Kconfig file. It appears only in the build-files. Whatever I set, it is always 1 for tinyDTLS.

Might be a bug when translating the Kconfig option. This is done [here](https://github.com/RIOT-OS/RIOT/blob/ffebb5d7929126581aeb960c881bba2f4f9754b8/pkg/tinydtls/Makefile.include#L69-L72). Have you tried setting it manually when building e.g.:

```
CFLAGS="-DDTLS_PEER_MAX=2" make -C examples/dtls-sock
```

janosbrodbeck commented on sys/include/net/gcoap/dtls_session_management.h:
```diff
@@ -0,0 +1,104 @@
+ * @return  1, on success
+ * @return  -1, when no existing session found
+ */
+ssize_t dtls_session_remove(sock_dtls_session_t *session);
```
> I don't see gcoap specific code in the `dtls_session_management` module other than reading the max peer config option. Maybe we could move it out of gcoap and make it an optional module for sock dtls instead?

I thought about that too. Despite the max peer config (which can be solved in a different way) there is nothing which tieds this module to gcoap exclusively. I move it out of gcoap.

The management itself definitely needs optimization (especially for higher session numbers), but that can be easily done later.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -128,14 +150,111 @@ static void *_event_loop(void *arg)
+            sock_dtls_session_destroy(sock, &session);
+        }
+
+        /* if no session slot is left: set timeout to free up session */
```
What's the reason for not calling `_dtls_free_up_session()` directly?

janosbrodbeck replied:
If more than one session is allowed yes: than it would be okay. But if only one session is allowed you don't want to destroy the session immediately after it was established. some time for coap messaging is needed.

benpicco reviewed (commented):
just some high-level comments

janosbrodbeck commented on sys/net/application_layer/gcoap/dtls_session_management.c:
```diff
@@ -0,0 +1,165 @@
+        if (restore) {
+            _ep_to_session(&session->ep, session);
+        }
+        session_slot->last_used = xtimer_now_usec64();
```
Depends on the application. A session can be kept for longer than an hour. In theory: a session has no expiration. Though I may make sense to drop it after longer period unused and redo the handshake.

janosbrodbeck commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -741,6 +889,11 @@ kernel_pid_t gcoap_init(void)
     }
+
+#ifdef CONFIG_GCOAP_ENABLE_DTLS
+    _main_thread = thread_getpid();
```
No, every thread can do CoAP communication. Now that I've looked at it again: This part should be moved to `_sock_send` and is not limited to the main_thread. 

--
To explain the idea behind this mechanic: Gcoap does only store the data that is to be sent on client side and when it is a confirmable message. It stores it in a memo which can be refound when the answer to the request arrives. 

If we want to send data and a session to the endpoint has not been established before we need to do a handshake first. Ideally you would initiate the handshake and when the handshake process is completed we send the data. If we want the handshake process to be asynchronous on gcoap side too, we need to store the data separate because, currently, we do not hold the data until the handshake is completed. 

Storing that data would cost `DTLS_HANDSHAKE_MAX` * `CONFIG_GCOAP_PDU_BUF_SIZE` (default 128 bytes) and we would need to attach this data somehow to the session (doable via session management I guess). `DTLS_HANDSHAKE_MAX` is the maximum of concurrent handshakes. 

That's a lot of additional memory and the reason why I decided to do it with blocking handshakes. When a session needs to be established we save the thread doing the handshake initiating and then wait till the process completes or timeouts.

When the successful handshake "arrives" in the DTLS event handler we inform the blocked thread that the handshake has been successful and completed -> `_sock_send` blocks until handshake is completed and can send data. We do not have to store additional data this way.

But it comes with a price - the gcoap server cannot do handshakes on its own this way. At the moment this is not needed. But in the meantime I have seen that it could(!) be necessary for observable resources. I have not yet fully understood a statement in the RFC standard regarding that.

janosbrodbeck commented:
After a long time finally a new push. This is my current WIP state. 

* Introduced `_tl_send()`, `_tl_authenticate()`, `_tl_get_socket()` functions as wrapper for UDP/DTLS
* Could not get rid of `#ifdef`/`#if` since the compiler did not optimize the DTLS functions out when not used with `IS_ACTIVE()` in case of not using DTLS. And without providing `USEPKG += tinydtls` the compiler just complains, that it cannot resolve the DTLS objects.
* Kconfig is addressed
* Session management (well, at the moment it is still more like session storage) as independent module for `sock_dtls`

The current state is still experimental. But all issues stem from the `sock_dtls` implementation of tinydtls. Because of that I've opened an issue to track those. It's a little difficult because this implementation is actually the only real user for `sock_dtls`/tinydtls and all the problems show up here, but it doesn't really have anything to do with gcoap. Because of that, there are currently two sock_dtls/tinydtls hotfix commits pushed. 

Known issues:
* On boards I have to send two messages, since the first authentication process always fails. Not reproducible on native. But could only test it on `particle-xenon` boards and over BLE. Would be nice, if someone else could test it on other boards. 

Still missing in gcoap/session management:
* Session expiration
* DTLS epoch check in the coap process

janosbrodbeck commented:
I've updated and rebased once again (mostly documentation). I remove the WIP state now. The issues with DTLS are not related to gcoap, nonetheless I keep the hotfix pkg/tinydtls commit for easier testing for now. Testing procedure from start post remains. I will also update the start post to the current state soon.

DTLS epoch checking and session expiration is not yet addressed, however, I would consider both to address in a follow-up.

Feedback and review is more than welcome.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -110,6 +126,17 @@ static event_queue_t _queue;
+/* DTLS variables and definitions */
+#define SOCK_DTLS_CLIENT_TAG (2)
+
+static sock_dtls_t _sock_dtls;
```
What if instead we did

```patch
--- a/pkg/tinydtls/include/sock_dtls_types.h
+++ b/pkg/tinydtls/include/sock_dtls_types.h
@@ -39,8 +39,8 @@ extern "C" {
  * @brief Information about DTLS sock
  */
 struct sock_dtls {
+    sock_udp_t udp_sock;                    /**< Underlying UDP sock to use */
     dtls_context_t *dtls_ctx;               /**< TinyDTLS context for sock */
-    sock_udp_t *udp_sock;                   /**< Underlying UDP sock to use */
 #if defined(SOCK_HAS_ASYNC) || defined(DOXYGEN)
     /**
      * @brief   Network stack internal buffer context.
```

And use the `.flags` field to indicate there is a DTLS context attached to the socket? 

Could that save the application from dealing with DTLS entirely (save for calling a DTLS setup function on the DTLS socket) and leave the rest to GNRC?

benpicco reviewed (commented):
Sorry for not taking a look at this earlier.
There are two conceptual things that irk me, but that might also be due to me not being familiar with the CoAP/DTLS/async socket implementation.

 - Why is DTLS sock not just an extension to UDP sock? The DTLS implementation would then be specific to the network stack (e.g. GNRC), but that would still be preferable to pushing it to the application layer IMHO
 - Why is there a need for `dtls/dsm` - can't we attach the session directly to the socket?

janosbrodbeck commented on sys/include/net/dsm.h:
```diff
@@ -0,0 +1,123 @@
+ * @return Previous state of the session. If no session existed before it returns
+ *         SESSION_STATE_NONE. If no space is available it returns NO_SPACE.
+ */
+dsm_state_t dsm_store(sock_dtls_t *sock, sock_dtls_session_t *session,
+                      dsm_state_t new_state, bool restore);
```
There can be `DTLS_PEER_MAX` sessions overall with e.g. tinyDTLS. How `DTLS_PEER_MAX` is distributed to the individual sockets then depends on actual usage. I could be that all peer slots are occupied for one socket, it can be evenly distributed, ...

janosbrodbeck replied:
Giving each DTLS socket `DTLS_PEER_MAX` session objects is extremely expensive. The other way would be to have a list of pointers of sessions. But then...you still need to store them somewhere.

janosbrodbeck commented:
>     * Why is there a need for `dtls/dsm` - can't we attach the session directly to the socket?
Like mentioned in one conversation comment, the distribution how the maximum number of sessions `DTLS_PEER_MAX` is distributed to individual sockets is unknown. In some place, currently existing sessions must be held.  Keeping actual sessions could of course be done in the implementation of the respective DTLS library. My decision here was to provide a simple generic way to hold sessions without each application/library implementation having to find its own solution.

Additionally, I have a scenario in mind: cross-application session management. Possibly a bit too far thought, but I can definitely see problems if an application can theoretically "occupy" all sessions indefinitely or for a very long time. By this I mean that e.g. `DTLS_PEER_MAX` sessions are occupied by gcoap. As a result, all other applications using DTLS have no possibility to establish connections or to receive incoming connections (kind of resource deadlock) on their sockets.

Right now dsm does not help here, but is very easy to extend to get access to sessions stored by others and close them to free peer resources. Provided dsm is used for storage by the respective applications.

Although I have built in the mechanism with gcoap for this very reason that sessions are automatically freed, but I consider the basic scenario with multiple DTLS applications far from impossible.

miri64 commented:
>     * Why is DTLS sock not just an extension to UDP sock? The DTLS implementation would then be specific to the network stack (e.g. GNRC), but that would still be preferable to pushing it to the application layer IMHO

As TLS is not just an extension of TCP (it is used with QUIC as well e.g.), DTLS is not an extension of UDP, as such you might want to have the possibility to use other datagram-based transports with DTLS. That's why DTLS sock is not just an extension to UDP sock.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+
+    uint8_t minimum_free = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS;
+    if (dsm_get_num_available_slots() < minimum_free) {
+        if(dsm_get_oldest_used_session(&_sock_dtls, &session) != -1) {
```
since this is the only use of this function, why not make it `dsm_drop_oldest_used_session()` and save yourself the need for iterating the sessions array twice?

janosbrodbeck replied:
You mean: Let `dsm` destroy the session instead of getting it and then destroy?

benpicco replied:
Yes, if the only reason to get the oldest session is to destroy it, it might as well be done internally.

janosbrodbeck replied:
My concern is that it moves a decision from the user to dsm. `sock_dtls_session_destroy()` is one part of the DTLS API which needs a refactor. When ignoring my currently applied tinyDTLS hotfix: `sock_dtls_session_destroy()` can fail. It merely initiates session destroy. But how do handle closing failure? Re-try? Force destroy? I'd argue that should decide the user per application. More details on this topic is in my PR for this: https://github.com/RIOT-OS/RIOT/pull/16422

That's why I avoided implementing any real session handling that concerns the DTLS stack in dsm.

benpicco commented on sys/net/dsm/dsm.c:
```diff
@@ -0,0 +1,176 @@
+    int res = -1;
+    dsm_session_t *session_slot = NULL;
+
+    if (dsm_get_num_available_slots() != DTLS_PEER_MAX) {
```
```suggestion
    if (dsm_get_num_available_slots() == DTLS_PEER_MAX) {
        return -1;
    }
```

You can drop a level of nesting

janosbrodbeck replied:
Done.

benpicco commented on sys/net/dsm/dsm.c:
```diff
@@ -0,0 +1,176 @@
+    if (empty_session) {
+        *session = empty_session;
+        return 0;
+    } else {
+        (void)session;
+    }
```
This can be dropped.

janosbrodbeck replied:
Dropped.

benpicco commented on sys/include/net/gcoap.h:
```diff
@@ -400,8 +420,31 @@ extern "C" {
 /**
  * @brief   Server port; use RFC 7252 default if not defined
  */
-#ifndef CONFIG_GCOAP_PORT
 #define CONFIG_GCOAP_PORT              (5683)
```
leave the `#ifndef` in place, add a new one for `CONFIG_GCOAPS_PORT`

janosbrodbeck replied:
Done.

benpicco commented on examples/gcoap/Makefile:
```diff
@@ -35,6 +35,23 @@ DEVELHELP ?= 1
+	# Maximum number of DTLS sessions
+	CFLAGS += -DDTLS_PEER_MAX=1
+	# increase main stack
+	CFLAGS += -DTHREAD_STACKSIZE_MAIN=2*THREAD_STACKSIZE_DEFAULT
```
is it not enough to bump `GCOAP_STACK_SIZE`?

janosbrodbeck replied:
The requests are sent from the main thread. Back then, I always had to increase the main stack because it wasn't quite enough. Tested it again, and it seems that no increase is necessary anymore. But I lack boards to be able to test it more extensively. However, if it is not enough on certain boards, this should quickly become noticeable.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+        if (sock_dtls_create(&_sock_dtls, &_sock_udp,
+                            CREDMAN_TAG_EMPTY,
+                            SOCK_DTLS_1_2, SOCK_DTLS_SERVER) < 0) {
+            puts("gcoap: error creating DTLS sock");
```
```suggestion
            DEBUG("gcoap: error creating DTLS sock");
```

janosbrodbeck replied:
Done.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
 
+    if (IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)) {
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+        sock_dtls_init();
```
that's already called by `sys/auto_init/auto_init.c`

janosbrodbeck replied:
True, but I still get segmentation fault, if I remove this line and try to establish a session with gcoap. Need to debug here :/

janosbrodbeck replied:
Fixed, but now depends on https://github.com/RIOT-OS/RIOT/pull/16528

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+        sock_dtls_session_get_udp_ep(&socket.ctx_dtls_session, &ep);
+        _process_coap_pdu(&socket, &ep,  _listen_buf, res);
+    }
+}
+#endif
+
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
```
```suggestion
}

```

janosbrodbeck replied:
Not sure if I understood this correctly

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+        }
+    }
+}
+#endif
```
```suggestion
#endif /* CONFIG_GCOAP_ENABLE_DTLS */
```

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+
+    if (type & SOCK_ASYNC_CONN_FIN) {
+        if (sock_dtls_get_event_session(sock, &socket.ctx_dtls_session)) {
+            dsm_remove(sock, &socket.ctx_dtls_session);
```
No `sock_dtls_session_destroy()` here?

janosbrodbeck replied:
`SOCK_ASYNC_CONN_FIN` is called when a session is closed from the remote peer. In that case, the session is already in closed state from tinyDTLS at this point. Since no other implementation exists, no idea how they would behave here. Would expect the same.

benpicco replied:
Maybe add a comment that the session is already destroyed - everywhere else `dsm_remove()` and `sock_dtls_session_destroy()` always appear together (so I was wondering if it is justified to move `sock_dtls_session_destroy()` into `dsm_remove()`)

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+        } else if (prev_state == NO_SPACE) {
+            /* No space in session management. Should not happen. If it occurs,
+            we lost track of sessions stored in tinydtls internally */
+            sock_dtls_session_destroy(sock, &socket.ctx_dtls_session);
```
if it shouldn't happen, maybe add some debug output / `assert(false)`

janosbrodbeck replied:
Added debug output here. Not sure whether `assert(false)` is really necessary here.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+        sock_udp_ep_t ep;
+        sock_dtls_session_get_udp_ep(&socket.ctx_dtls_session, &ep);
+
+        /* Remove all memos of the concerned session. TODO: oberservable memos! */
+        for (int i = 0; i < CONFIG_GCOAP_REQ_WAITING_MAX; i++) {
+            if (_coap_state.open_reqs[i].state == GCOAP_MEMO_UNUSED) {
```
Do we need to lock `_coap_state.lock` here?

janosbrodbeck replied:
Not 100% sure, from looking into code I didn't see locking when memos were set to unused. Since no moving or overwriting happens and no other* thread could access this memo I would say it's safe without.

* This path occurs when the remote peer closed the session, so we are already in the receiver thread. If context switch happens to a sending thread, we would only create a new memo and not edit existing ones.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -299,7 +446,9 @@ static void _on_resp_timeout(void *arg) {
+        coap_socket_t socket;
+        _tl_get_socket(&socket);
+        ssize_t bytes = _tl_send(&socket, memo->msg.data.pdu_buf,
                                       memo->msg.data.pdu_len, &memo->remote_ep);
```
please align ;)

janosbrodbeck replied:
Done.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -732,6 +881,95 @@ static void _find_obs_memo_resource(gcoap_observe_memo_t **memo,
+ * Transport layer functions
+ */
+
+static void _tl_get_socket(coap_socket_t *sock)
```
```suggestion
static void _tl_init_socket(coap_socket_t *sock)
```

janosbrodbeck replied:
Wouldn't `tl_init_coap_socket()` be more clear?

janosbrodbeck replied:
Changed.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -732,6 +881,95 @@ static void _find_obs_memo_resource(gcoap_observe_memo_t **memo,
+                      const sock_udp_ep_t *remote)
+{
+    int res;
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
```
We should probably also check `sock->type` here so a DTLS enabled client can still send unencrypted requests

janosbrodbeck replied:
True. Theoretically right now, we could send unencrypted...but AFAIK we "cannot" receive unencrypted. (We can, but this would just lead to encryption/DTLS errors, since DTLS headers are expected)

benpicco replied:
Hm what would be needed to also being able to receive non-encrypted answers?
We could do that as a follow-up, but generally it should still be possible to do unencrypted requests when connecting to a non-COAPS server

janosbrodbeck replied:
1) gcoap API change to distinguish between chosen transport types
2) Another UDP socket for running the non-COAPS server.
That should be enough to make it work.

janosbrodbeck replied:
That would mean that we expect to communicate with RIOT nodes in the default configuration. 5684 may be the default for gcoap over DTLS, but it is not mandatory.

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -732,6 +881,95 @@ static void _find_obs_memo_resource(gcoap_observe_memo_t **memo,
+    return res;
+}
+
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+static ssize_t _tl_authenticate(coap_socket_t *sock, const sock_udp_ep_t *remote,
+                                uint32_t timeout)
+{
```
```suggestion
static ssize_t _tl_authenticate(coap_socket_t *sock, const sock_udp_ep_t *remote,
                                uint32_t timeout)
{
#if !IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
    return 0;
#else
```

janosbrodbeck replied:
Done with a single if

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -732,6 +881,95 @@ static void _find_obs_memo_resource(gcoap_observe_memo_t **memo,
+        uint32_t start = xtimer_now_usec();
+        res = xtimer_msg_receive_timeout(&msg, timeout);
+
+        if (timeout != SOCK_NO_TIMEOUT) {
+            uint32_t diff = (xtimer_now_usec() - start);
+            timeout = (diff > timeout) ? 0: timeout - diff;
+            is_timed_out = (res < 0) || (timeout == 0);
+        }
```
What is going on here?

janosbrodbeck replied:
This is the mechanism from sock_dtls for blocking handshake in `sock_dtls_send()`. For the case that we receive another msg than the expected in this time, we wait the full timeout time to receive `DTLS_CONNECT`.

benpicco replied:
Can you add a comment about this?
Future readers won't be hunting down the PR comment :wink:

benpicco commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -887,8 +1125,16 @@ size_t gcoap_req_send(const uint8_t *buf, size_t len,
+    coap_socket_t socket = { 0 };
+
+    _tl_get_socket(&socket);
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
```
You can drop the `#ifdef` with the change suggested above

janosbrodbeck replied:
Dropped, and checked  the socket type here.

janosbrodbeck commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -30,6 +30,27 @@
+#include "net/dsm.h"
+#include "tinydtls_keys.h"
+
+#define GCOAP_DTLS_CREDENTIAL_TAG 10
```
None. It's just a number I used here for the example. Credman does not allow storing multiple credentials with the same tag + type (PSK/RPK). The user would need to avoid same tag numbers in real applications.

benpicco replied:
better add a comment, since this will serve as an example

janosbrodbeck commented on sys/net/dsm/dsm.c:
```diff
@@ -0,0 +1,176 @@
+    sock_dtls_t *sock;
+    sock_dtls_session_t session;
+    dsm_state_t state;
+    uint64_t last_used;
```
Will add unit. ms accuracy is sufficient. With ms we would have an overflow after 18 days? Right now there is now session timeout, so theoretically a session could be kept longer. The standard recommends to time out sessions after a maximum of 24h hours. With 32 bit, wouldn't we need two 32 bit values then? One for `established_ms` and one for `last_used_ms`, for the case when the current runtime of the device is longer than 18 days?

benpicco replied:
Wouldn't it be enough if `last_used_ms` was > 24h for a timeout?
On those scales even second accuracy would probably be enough,

benpicco replied:
If old sessions get evicted automatically when space runs out, I don't think we need to remove them manually after a timeout - nothing to gain from that.

janosbrodbeck replied:
Ultimately, the timeout is again application-specific and can be relevant to security. So, any kind of mechanism would be nice here which does not rely on "only if our session storage is full".

benpicco commented on sys/net/dsm/dsm.c:
```diff
@@ -0,0 +1,176 @@
+    dsm_session_t *session_slot = NULL;
+    mutex_lock(&_lock);
+    if (_find_session(sock, session, &session_slot) == 1) {
+        memset(&session_slot->session, 0 , sizeof(session_slot->session));
```
```suggestion
        memset(&session_slot->session, 0, sizeof(session_slot->session));
```

janosbrodbeck replied:
Yes, removed the memset.

janosbrodbeck commented:
Addressed first comments. Additionally:
* Made one gcoap API change: Changed return type from `gcoap_req_send` from size_t to int, to allow to return negative values. In this case, `gcoap_req_send` now returns `-ENOTCONN`, when the authentication process with the remote peer failed.

miri64 commented:
>     * Made one gcoap API change: Changed return type from `gcoap_req_send` from size_t to int, to allow negative return values. In this case, `gcoap_req_send` now returns `-ENOTCONN`, when the authentication process with the remote peer failed.

Why not `ssize_t` then?

miri64 commented:
> > ```
> > * Made one gcoap API change: Changed return type from `gcoap_req_send` from size_t to int, to allow negative return values. In this case, `gcoap_req_send` now returns `-ENOTCONN`, when the authentication process with the remote peer failed.
> > ```
> 
> 
> Why not `ssize_t` then?

And should this maybe moved to another PR, so the API change is more prominent?

janosbrodbeck commented:
> > ```
> > * Made one gcoap API change: Changed return type from `gcoap_req_send` from size_t to int, to allow negative return values. In this case, `gcoap_req_send` now returns `-ENOTCONN`, when the authentication process with the remote peer failed.
> > ```
> 
> Why not `ssize_t` then?
True, I change it to ssize_t, that makes sense to keep, because positive values still returns sent bytes.

Regarding moving to a separate PR: I can do that. Gives possibly also again more attention that this PR is (IMHO) in advanced state.

miri64 commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -316,7 +338,16 @@ int gcoap_cli_cmd(int argc, char **argv)
+        }
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+            printf("Connection secured with DTLS\n");
+            printf("Free DTLS session slots: %d/%d\n", dsm_get_num_available_slots(),
```
Indentation is off here

janosbrodbeck replied:
Addressed with the newest commits.

miri64 commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -466,5 +497,19 @@ int gcoap_cli_cmd(int argc, char **argv)
 void gcoap_cli_init(void)
 {
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+        int res = credman_add(&credential);
+        if (res < 0 && res != CREDMAN_EXIST) {
+            /* ignore duplicate credentials */
+            printf("gcoap: cannot add credential to system: %d\n", res);
+            return;
+        }
+        sock_dtls_t *gcoap_sock_dtls = gcoap_get_sock_dtls();
+        res = sock_dtls_add_credential(gcoap_sock_dtls, GCOAP_DTLS_CREDENTIAL_TAG);
+        if (res < 0) {
+            printf("gcoap: cannot add credential to DTLS sock: %d\n", res);
+        }
```
Here too.

janosbrodbeck replied:
Addressed with the newest commits.

miri64 reviewed (commented):
Just some style nits because I just noticed them

janosbrodbeck commented:
Will revert/adjust the `gcoap_req_send()` change when/if https://github.com/RIOT-OS/RIOT/pull/16526 is merged. Have to rebase then anyway.

miri64 commented:
> Will revert/adjust the `gcoap_req_send()` return type change when/if #16526 is merged. Have to rebase then anyway.

#16526 is merged :-).

janosbrodbeck commented:
Rebased  to upstream.

janosbrodbeck commented:
Murdock is unhappy because it does not fit anymore on some boards. That was to be expected for less-good equipped boards. The default should still be non-DTLS. A separate test for gcoap-dtls is still needed, otherwise too much code would not be tested by the builds.

benpicco commented:
probably best to create `examples/gcoap-dtls` as a copy of `examples/gcoap` then.
I'm not sure what's the best approach there as `gcoap_cli.c` would create a lot of copy & paste between the two.
Maybe those are generic enough to be moved to `sys/shell/commands` ?

miri64 commented:
> I'm not sure what's the best approach there as `gcoap_cli.c` would create a lot of copy & paste between the two.

If the files are identical, one could just use symlinks. They are supported by Git.

janosbrodbeck commented:
> > I'm not sure what's the best approach there as `gcoap_cli.c` would create a lot of copy & paste between the two.
> 
> If the files are identical, one could just use symlinks. They are supported by Git.

Not 100% identical, there are small additions. Would it be an option to keep the additions in `examples/gcoap` and only symlink `gcoap_cli.c` for `examples/gcoap-dtls`? A makefile configured for DTLS and the keys would be enough in this example.

janosbrodbeck commented:
Added a seperate gcoap-dtls example to see what murdock thinks about this.

benpicco commented on sys/include/net/gcoap.h:
```diff
@@ -384,6 +401,9 @@
 #include "event/timeout.h"
 #include "net/ipv6/addr.h"
 #include "net/sock/udp.h"
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
```
Why not 
```suggestion
#if IS_ACTIVE(MODULE_GCOAP_DTLS)
```

miri64 replied:
`s/IS_ACTIVE/IS_USED/` ;-)

benpicco replied:
Given that we have

```C
#define IS_USED(module) IS_ACTIVE(module)
```

we should just get rid of either define - but that's unrelated to this PR

janosbrodbeck replied:
Is there any preference regarding the two macros? For modules `IS_USED` sounds a bit more appropriate

benpicco commented on sys/include/net/gcoap.h:
```diff
@@ -403,6 +423,33 @@ extern "C" {
 #ifndef CONFIG_GCOAP_PORT
 #define CONFIG_GCOAP_PORT              (5683)
 #endif
+#ifndef CONFIG_GCOAPS_PORT
```
```suggestion
/**
 * @brief   Secure Server port; use RFC 7252 default if not defined
 */
#ifndef CONFIG_GCOAPS_PORT
```

janosbrodbeck replied:
Done. Also documented the `coap_socket_t` struct now.

benpicco commented on sys/include/net/gcoap.h:
```diff
@@ -567,9 +614,14 @@ extern "C" {
  */
 #ifndef GCOAP_STACK_SIZE
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+#define GCOAP_STACK_SIZE (2*THREAD_STACKSIZE_DEFAULT + DEBUG_EXTRA_STACKSIZE \
```
let's rather define a `DTLS_EXTRA_STACKSIZE` that gets added here

janosbrodbeck replied:
Done. I always define `GCOAP_DTLS_EXTRA_STACKSIZE`, but set it to 0 for non-DTLS.

benpicco commented on sys/include/net/gcoap.h:
```diff
@@ -719,6 +772,31 @@ typedef struct {
+    coap_socket_type_t type;
+    union {
+        sock_udp_t *udp;
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS) || defined(DOXYGEN)
```
also

```suggestion
#if IS_ACTIVE(MODULE_GCOAP_DTLS) || defined(DOXYGEN)
```

janosbrodbeck replied:
True. Replaced `CONFIG_GCOAP_ENABLE_DTLS` completely with `IS_USED(MODULE_GCOAP_DTLS)`

benpicco commented on sys/net/dsm/dsm.c:
```diff
@@ -0,0 +1,171 @@
+
+    mutex_lock(&_lock);
+    for (uint8_t i=0; i < DTLS_PEER_MAX; i++) {
+        if (_sessions[i].state == SESSION_STATE_ESTABLISHED
+                && _sessions[i].sock == sock) {
+            if (session_slot == NULL
+                || session_slot->last_used_sec > _sessions[i].last_used_sec)
+            {
+                session_slot = &_sessions[i];
+            }
+        }
```
style nit: it's possible to reduce the level of nesting here

```suggestion
        if (_sessions[i].state != SESSION_STATE_ESTABLISHED) {
            continue;
        }
        if (_sessions[i].sock != sock) {
            continue;
        }

        if (session_slot == NULL ||
            session_slot->last_used_sec > _sessions[i].last_used_sec) {
            session_slot = &_sessions[i];
        }
```

janosbrodbeck replied:
Done.

benpicco commented on sys/net/dsm/dsm.c:
```diff
@@ -0,0 +1,171 @@
+    mutex_lock(&_lock);
+
+    ssize_t res = _find_session(sock, session, &session_slot);
+    if (res != -1) {
```
```suggestion
    if (res >= 0) {
```

adds a bit of future-proofing should `_find_session` ever get refactored to emit more error code

or even better

```suggestion
    if (res < 0) {
        DEBUG("dsm: no space for session to store\n");
        goto out;
    }
```

and we can reduce the level of nesting

janosbrodbeck replied:
Done.

benpicco reviewed (commented):
CI still has some [style nits](https://github.com/RIOT-OS/RIOT/pull/15549/files) so I added my own

janosbrodbeck commented:
I've also renamed `dsm_get_oldest_used_session()` to `dsm_get_least_recently_used_session()` to get rid of this awful naming. Though about using LRU to shorten it, but nah, I think it's more clear so even if it's longer.

benpicco commented:
Please squash (& rebase)

janosbrodbeck commented:
> Please squash (& rebase)

Squashed and rebased. Also removed the tinyDTLS workaround commit, since it is now merged into master.
There is one change compared to before. I've added the following marked lines:
https://github.com/janosbrodbeck/RIOT/commit/31518c7a4f62cd420e5ba323052bc48115b3a418#diff-210ffdc67fd874379abf19f95a95fe1f6d7c465aa69e20b6e4fcfd625fbcae18R94-R100

Without this, it could happen that two session slots were added as available for one closed session.

benpicco commented on sys/net/dsm/dsm.c:
```diff
@@ -0,0 +1,184 @@
+#include "net/sock/util.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG 1
```
```suggestion
#define ENABLE_DEBUG 0
```

janosbrodbeck replied:
Urgh... I thought about it just before pushing and then forgot about it

benpicco reviewed (commented):
I think there is still something wrong here:
When playing around with two native RIOT nodes querying a third native RIOT node with

    coap get fe80::7837:fcff:fe7d:1aaf 5684 /.well-known/core

after a while one of the nodes would no longer receive a response. (This is after the `dsm: removed session` message), the server shows

```
CoAP server is listening on port 5684
Connection secured with DTLS
Free DTLS session slots: 1/1
 CLI requests sent: 0
CoAP open requests: 0
Configured Proxy: None
```

The clients meanwhile display

```
CoAP server is listening on port 5684
Connection secured with DTLS
Free DTLS session slots: 13/1
 CLI requests sent: 17
CoAP open requests: 0
Configured Proxy: None
```

This number increases with each request that does not receive a response.

janosbrodbeck commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -119,23 +146,139 @@ static void *_event_loop(void *arg)
+
+    uint8_t minimum_free = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS;
+    if (dsm_get_num_available_slots() < minimum_free) {
+        if(dsm_get_least_recently_used_session(&_sock_dtls, &session) != -1) {
```
> I think there is still something wrong here:
> When playing around with two native RIOT nodes querying a third native RIOT node with
> 
> ```
> coap get fe80::7837:fcff:fe7d:1aaf 5684 /.well-known/core
> ```
> 
> after a while one of the nodes would no longer receive a response. (This is after the `dsm: removed session` message), the server shows
> 
> ```
> CoAP server is listening on port 5684
> Connection secured with DTLS
> Free DTLS session slots: 1/1
>  CLI requests sent: 0
> CoAP open requests: 0
> Configured Proxy: None
> ```
> 
> The clients meanwhile display
> 
> ```
> CoAP server is listening on port 5684
> Connection secured with DTLS
> Free DTLS session slots: 13/1
>  CLI requests sent: 17
> CoAP open requests: 0
> Configured Proxy: None
> ```
> 
> The number of 'free sessions' increases with each request that does not receive a response.

Fixed. It was a problem with finding sessions which have been removed but stated them as "already existing", since session state was not taken into account.

janosbrodbeck commented:
> I think there is still something wrong here:
> The number of 'free sessions' increases with each request that does not receive a response.

Fixed. It was a problem with finding sessions which have been removed but stated them as "already existing", since session state was not taken into account.

benpicco commented:
Thank you, works now!
Please squash.

janosbrodbeck commented:
Squashed.

benpicco commented:
Thank you!

janosbrodbeck commented:
Thanks for the help and review, everyone!

fjmolinas commented:
> I think there is still something wrong here: When playing around with two native RIOT nodes querying a third native RIOT node with
> 
> ```
> coap get fe80::7837:fcff:fe7d:1aaf 5684 /.well-known/core
> ```
> 
> after a while one of the nodes would no longer receive a response. (This is after the `dsm: removed session` message), the server shows
> 
> ```
> CoAP server is listening on port 5684
> Connection secured with DTLS
> Free DTLS session slots: 1/1
>  CLI requests sent: 0
> CoAP open requests: 0
> Configured Proxy: None
> ```
> 
> The clients meanwhile display
> 
> ```
> CoAP server is listening on port 5684
> Connection secured with DTLS
> Free DTLS session slots: 13/1
>  CLI requests sent: 17
> CoAP open requests: 0
> Configured Proxy: None
> ```
> 
> The number of 'free sessions' increases with each request that does not receive a response.

I was trying to test #17141, and was for this using the `examples/gcoap_dtls` application, for some reason I could not get it to work between real BOARDS and with native I could, I went back to this PR to double check and I experience the same issue, might I be missing something? I'm flashing the cpplication to 2 BOARDS and the performing a get on `/.well-known/core`
